packages/grpc-reflection/src/implementations/reflection-v1.ts

```diff
@@ -0,0 +1,334 @@
+    const pkg = grpc.loadPackageDefinition(pkgDefinition) as any;
+
+    server.addService(pkg.grpc.reflection.v1.ServerReflection.service, {
+      ServerReflectionInfo: (
+        stream: grpc.ServerDuplexStream<ServerReflectionRequest, ServerReflectionResponse>
+      ) => {
+        stream.on('end', () => stream.end());
+
+        stream.on('data', (message: ServerReflectionRequest) => {
+          stream.write(this.handleServerReflectionRequest(message));
+        });
+      }
+    });
+  }
+
+  /** Assemble a response for a single server reflection request in the stream */
+  handleServerReflectionRequest(message: ServerReflectionRequest): ServerReflectionResponse {
+    const response: ServerReflectionResponse = {
+      validHost: message.host,
+      originalRequest: message,
+      fileDescriptorResponse: undefined,
+      allExtensionNumbersResponse: undefined,
+      listServicesResponse: undefined,
+      errorResponse: undefined,
+    };
+
+    try {
+      if (message.listServices !== undefined) {
+        response.listServicesResponse = this.listServices(message.listServices);
+      } else if (message.fileContainingSymbol !== undefined) {
+        response.fileDescriptorResponse = this.fileContainingSymbol(message.fileContainingSymbol);
+      } else if (message.fileByFilename !== undefined) {
+        response.fileDescriptorResponse = this.fileByFilename(message.fileByFilename);
+      } else if (message.fileContainingExtension !== undefined) {
+        response.fileDescriptorResponse = this.fileContainingExtension(
+          message.fileContainingExtension?.containingType || '',
+          message.fileContainingExtension?.extensionNumber || -1
+        );
+      } else if (message.allExtensionNumbersOfType) {
+        response.allExtensionNumbersResponse = this.allExtensionNumbersOfType(message.allExtensionNumbersOfType);
+      } else {
+        throw new ReflectionError(
+          grpc.status.UNIMPLEMENTED,
+          `Unimplemented method for request: ${message}`,
+        );
+      }
+    } catch (e) {
+      if (e instanceof ReflectionError) {
+        response.errorResponse = {
+          errorCode: e.statusCode,
+          errorMessage: e.message,
+        };
+      } else {
+        response.errorResponse = {
+          errorCode: grpc.status.UNKNOWN,
+          errorMessage: 'Failed to process gRPC reflection request: unknown error',
+        };
+      }
+    }
+
+    return response;
+  }
+
+  /** List the full names of registered gRPC services
+   *
+   * note: the spec is unclear as to what the 'listServices' param can be; most
+   * clients seem to only pass '*' but unsure if this should behave like a
+   * filter. Until we know how this should behave with different inputs this
+   * just always returns *all* services.
+   *
+   * @returns full-qualified service names (eg. 'sample.SampleService')
+   */
+  listServices(listServices: string): ListServiceResponse__Output {
```

Comment:
it can as-written but tbh I'm not 100% clear on what the input to this method is intended to be (you'll notice it's currently unused) so it might actually be relevant to the function and need to be used here

the [proto docs for this method](https://github.com/grpc/grpc/blob/master/src/proto/grpc/reflection/v1/reflection.proto#L69) are unclear as to whether this is supposed to be a filter or something else, but I've noticed that some of the clients in testing use `"*"` and it doesn't hurt to return all so that's just the behavior always at the moment

Reply:
The comment on that field says "The content will not be checked." I would interpret that to mean that the server shouldn't do anything with it.

Reply:
ah yeah that makes sense - done